coverage/summary.py

```diff
@@ -30,6 +30,112 @@ def writeout(self, line):
+            Missing="{:>9}")
+        header_items = [
+            h_form[item].format(item, name_len=max_name, n=max_n)
+            for item in header]
+        header_str = "".join(header_items)
+        rule = "-" * len(header_str)
+
+        # Write the header
+        self.writeout(header_str)
+        self.writeout(rule)
+
+        column_order = dict(name=0, stmts=1, miss=2, cover=-1)
+        if self.branches:
+            column_order.update(dict(branch=3, brpart=4))
+
+        h_form.update(dict(Cover="{:>{n}}%"), Missing=" {:9}")
+        for values in lines_values:
+            # build string with line values
+            line_items = [
+                h_form[item].format(str(value),
+                name_len=max_name, n=max_n-1) for item, value in zip(header, values)]
+            text = "".join(line_items)
+            self.writeout(text)
+
+        # Write a TOTAL line
+        if total_line:
+            self.writeout(rule)
+            line_items = [
+                h_form[item].format(str(value),
+                name_len=max_name, n=max_n-1) for item, value in zip(header, total_line)]
+            text = "".join(line_items)
+            self.writeout(text)
+
+        for end_line in end_lines:
+            self.writeout(end_line)
+        return self.total.n_statements and self.total.pc_covered
+
+    def _report_markdown(self, header, lines_values, total_line, end_lines):
+        "internal method to print report data in markdown format"
+        # Prepare the formatting strings, header, and column sorting.
+        max_name = max([len(fr.relative_filename().replace("_","\\_")) for\
+            (fr, analysis) in self.fr_analysis] + [9]) + 1
+        h_form = dict(
+            Name="| {:{name_len}}|", Stmts="{:>7} |", Miss="{:>7} |",
+            Branch="{:>7} |", BrPart="{:>7} |", Cover="{:>{n}} |",
+            Missing="{:>9} |")
+        n = self.config.precision
+        max_n = max(n+6, 7) + 4
+        header_items = [
+            h_form[item].format(item, name_len=max_name, n=max_n) for item in header]
+        header_str = "".join(header_items)
+        rule_str = "|" + " ".join(["- |".rjust(len(header_items[0])-1, '-')] +
+            ["-: |".rjust(len(item)-1, '-') for item in header_items[1:]])
+
+        # Write the header
+        self.writeout(header_str)
+        self.writeout(rule_str)
+
+        column_order = dict(name=0, stmts=1, miss=2, cover=-1)
+        if self.branches:
+            column_order.update(dict(branch=3, brpart=4))
+
+        for values in lines_values:
+            # build string with line values
+            h_form.update(dict(Cover="{:>{n}}% |"))
+            line_items = [
+                h_form[item].format(str(value).replace("_", "\\_"),
+                name_len=max_name, n=max_n-1) for item, value in zip(header, values)]
+            text = "".join(line_items)
+            self.writeout(text)
+
+        # Write the TOTAL line
+        if total_line:
```

Comment:
I ran a coverage report locally, and this line (and the similar one in `_report_text`) say the condition is always true.  Is it true that these functions can never be called with an empty `total_line`, or do we just not have a test for that case?

Reply:
The total line is always printed, if there is coverage data to report.

The `report` method raises an Exception before creating the total line, if there is no coverage data to report.